File: apps/android/lib/src/domain/sync_manifest.dart
/// Dart mirrors of the srv-32 sync-manifest contract
/// (`server/src/workspace/sync-manifest.ts`, plan 191). Pure data with no
/// platform dependencies, so the delta-sync planner can be unit-tested against
/// them without any IO. The companion diffs this two-level manifest to pull
/// only changed chapters.
library;

/// One lightweight book row from the manifest INDEX.
class SyncManifestIndexBook {
  const SyncManifestIndexBook({
    required this.bookId,
    required this.updatedAt,
    required this.title,
    required this.author,
    required this.series,
    required this.seriesPosition,
    required this.chapterCount,
    this.coverUrl,
    this.finished = false,
    this.hidden = false,
  });

  final String bookId;

  /// Audio-aware ISO timestamp — moves on a metadata edit OR an audio regen.
  /// ISO-8601 UTC strings compare chronologically, so a `?since` diff is a
  /// plain string comparison.
  final String updatedAt;
  final String title;
  final String author;
  final String series;
  final double? seriesPosition;

  /// Active (non-excluded) chapter count.
  final int chapterCount;
  final String? coverUrl;
  final bool finished;
  final bool hidden;

  factory SyncManifestIndexBook.fromJson(Map<String, dynamic> json) {
    return SyncManifestIndexBook(
      bookId: json['bookId'] as String,
      updatedAt: json['updatedAt'] as String,
      title: json['title'] as String? ?? '',
      author: json['author'] as String? ?? '',
      series: json['series'] as String? ?? '',
      seriesPosition: (json['seriesPosition'] as num?)?.toDouble(),
      chapterCount: (json['chapterCount'] as num?)?.toInt() ?? 0,
      coverUrl: json['coverUrl'] as String?,
      finished: json['finished'] as bool? ?? false,
      hidden: json['hidden'] as bool? ?? false,
    );
  }
}

/// The manifest INDEX: one row per book plus the full active-book set, which
/// drives stateless client-side eviction (a filesystem scan has no tombstones).
class SyncManifestIndex {
  const SyncManifestIndex({
    required this.schemaVersion,
    required this.books,
    required this.activeBookIds,
  });

  final int schemaVersion;
  final List<SyncManifestIndexBook> books;

  /// ALWAYS the full current set, even under `?since` — the client evicts any
  /// local book absent from it.
  final List<String> activeBookIds;

  factory SyncManifestIndex.fromJson(Map<String, dynamic> json) {
    final books = (json['books'] as List<dynamic>? ?? const [])
        .map((e) => SyncManifestIndexBook.fromJson(e as Map<String, dynamic>))
        .toList(growable: false);
    final active = (json['activeBookIds'] as List<dynamic>? ?? const [])
        .map((e) => e as String)
        .toList(growable: false);
    return SyncManifestIndex(
      schemaVersion: (json['schemaVersion'] as num?)?.toInt() ?? 0,
      books: books,
      activeBookIds: active,
    );
  }
}

/// One chapter from a book's manifest DETAIL, keyed by the stable srv-35
/// `uuid`. `fingerprint`/`urlSuffix`/`audioUrl` are absent when the chapter has
/// no rendered audio.
class SyncManifestChapter {
  const SyncManifestChapter({
    required this.uuid,
    required this.id,
    required this.title,
    this.fingerprint,
    this.urlSuffix,
    this.audioUrl,
    this.durationSec,
    this.lufs,
  });

  /// Stable identifier (srv-35) — the key for sync + bookmarks. NEVER use the
  /// positional [id] as a key.
  final String uuid;

  /// Current positional id — only used to build [audioUrl] server-side; the
  /// client already receives the built [audioUrl].
  final int id;
  final String title;

  /// `audioRenderedAt|fileSize` — changes on every audio-mutating server path.
  final String? fingerprint;

  /// The actual rendered format: `audio.mp3` | `audio.m4a` | `audio.ogg`.
  /// Never hardcode `.mp3`.
  final String? urlSuffix;
  final String? audioUrl;
  final double? durationSec;
  final double? lufs;

  bool get hasAudio => audioUrl != null && fingerprint != null;

  /// The expected on-disk byte size, parsed from the fingerprint's size
  /// component, used for the post-download integrity check. Null without audio.
  int? get expectedSize {
    final fp = fingerprint;
    if (fp == null) return null;
    final parts = fp.split('|');
    if (parts.length != 2) return null;
    return int.tryParse(parts[1]);
  }

  factory SyncManifestChapter.fromJson(Map<String, dynamic> json) {
    return SyncManifestChapter(
      uuid: json['uuid'] as String,
      id: (json['id'] as num).toInt(),
      title: json['title'] as String? ?? '',
      fingerprint: json['fingerprint'] as String?,
      urlSuffix: json['urlSuffix'] as String?,
      audioUrl: json['audioUrl'] as String?,
      durationSec: (json['durationSec'] as num?)?.toDouble(),
      lufs: (json['lufs'] as num?)?.toDouble(),
    );
  }
}

/// A single book's manifest DETAIL: its active chapters (keyed by `uuid`) plus
/// the full active-chapter set for client-side eviction.
class SyncManifestBookDetail {
  const SyncManifestBookDetail({
    required this.schemaVersion,
    required this.bookId,
    required this.updatedAt,
    required this.chapters,
    required this.activeChapterUuids,
  });

  final int schemaVersion;
  final String bookId;
  final String updatedAt;
  final List<SyncManifestChapter> chapters;

  /// Full active-chapter set for this book — the client evicts any local
  /// chapter absent from it.
  final List<String> activeChapterUuids;

  factory SyncManifestBookDetail.fromJson(Map<String, dynamic> json) {
    final chapters = (json['chapters'] as List<dynamic>? ?? const [])
        .map((e) => SyncManifestChapter.fromJson(e as Map<String, dynamic>))
        .toList(growable: false);
    final active = (json['activeChapterUuids'] as List<dynamic>? ?? const [])
        .map((e) => e as String)
        .toList(growable: false);
    return SyncManifestBookDetail(
      schemaVersion: (json['schemaVersion'] as num?)?.toInt() ?? 0,
      bookId: json['bookId'] as String,
      updatedAt: json['updatedAt'] as String,
      chapters: chapters,
      activeChapterUuids: active,
    );
  }
}

File: server/src/workspace/sync-manifest.ts
/* srv-32 (plan 191) — pure builders for the companion sync manifest.
 *
 * A two-level manifest the Android companion diffs to delta-sync a
 * constantly-regenerated library:
 *
 *  - INDEX (`buildSyncManifestIndex`): one lightweight row per book
 *    (bookId + audio-aware `updatedAt` + cover ref + active-id set). A
 *    `?since` cutoff trims the `books` list but NEVER the `activeBookIds`
 *    set — that full set drives stateless client-side deletion (a
 *    filesystem scan has no tombstones for removed books).
 *  - DETAIL (`buildSyncManifestBookDetail`): one book's chapters keyed by
 *    the stable srv-35 `uuid`, each with a fingerprint (`audioRenderedAt`
 *    + file size — so any audio mutation bumps it), the actual rendered
 *    `urlSuffix`/`audioUrl`, and duration/lufs hints.
 *
 * Pure — all disk I/O (book walk, file stat, format probe) happens in the
 * route (library-sync-manifest.ts) and is passed in. See plan 191. */

import type { BookStateJson } from './scan.js';

/** Bump on any breaking change to the manifest shape. Surfaced in
 *  `GET /api/info` `schemas.syncManifest` so the companion can compat-gate. */
export const SYNC_MANIFEST_SCHEMA = 1;

/** Per-chapter on-disk audio facts the route gathers (stat + format probe)
 *  and hands to the detail builder. Absent from the map for a chapter with
 *  no rendered audio. */
export interface ChapterAudioFact {
  fileSize: number;
  urlSuffix: 'audio.mp3' | 'audio.m4a' | 'audio.ogg';
  /** Real PCM-measured length from the chapter's `<slug>.segments.json`
   *  (the authoritative source the library scan uses). Preferred over the
   *  QA verdict's `durationSec`, which is often absent on older books. */
  durationSec?: number;
}

export interface SyncManifestIndexBook {
  bookId: string;
  updatedAt: string;
  title: string;
  author: string;
  series: string;
  seriesPosition: number | null;
  /** Active (non-excluded) chapter count. */
  chapterCount: number;
  coverUrl?: string;
  /** Explicit "Mark as finished" flag from listen-progress.json (NOT the
   *  derived isFinished — durations aren't reliably loaded on the index path). */
  finished?: boolean;
  /** "Hide from shelf" flag from listen-progress.json. */
  hidden?: boolean;
}

export interface SyncManifestIndex {
  schemaVersion: number;
  books: SyncManifestIndexBook[];
  /** ALWAYS the full current set, even under `?since` — the client evicts
   *  any local book absent from it. */
  activeBookIds: string[];
}

export interface SyncManifestChapter {
  uuid: string;
  /** Current positional id — the client builds `audioUrl` from it; keying
   *  is by `uuid`. */
  id: number;
  title: string;
  /** `audioRenderedAt` + file size; absent when the chapter has no audio. */
  fingerprint?: string;
  urlSuffix?: 'audio.mp3' | 'audio.m4a' | 'audio.ogg';
  audioUrl?: string;
  durationSec?: number;
  lufs?: number;
}

export interface SyncManifestBookDetail {
  schemaVersion: number;
  bookId: string;
  updatedAt: string;
  chapters: SyncManifestChapter[];
  /** Full active-chapter set for this book — the client evicts any local
   *  chapter absent from it. */
  activeChapterUuids: string[];
}

/** A chapter's content fingerprint: changes whenever the rendered audio
 *  changes, because both inputs (the render timestamp and the byte size)
 *  move on every audio-mutating path (all converge on
 *  finalize-chapter-write.ts, which stamps a fresh `audioRenderedAt`).
 *  Undefined when the chapter has no rendered audio. */
export function chapterFingerprint(
  audioRenderedAt: string | undefined,
  fileSize: number | undefined,
): string | undefined {
  if (!audioRenderedAt || fileSize === undefined) return undefined;
  return `${audioRenderedAt}|${fileSize}`;
}

/** ISO-comparable max of `state.updatedAt` and every chapter's
 *  `audioRenderedAt`, so the index signal moves on a metadata edit OR an
 *  audio regen. (ISO-8601 UTC strings compare chronologically.) */
export function bookManifestUpdatedAt(state: BookStateJson): string {
  let max = state.updatedAt;
  for (const c of state.chapters) {
    if (c.audioRenderedAt && c.audioRenderedAt > max) max = c.audioRenderedAt;
  }
  return max;
}

export function buildSyncManifestIndex(
  books: ReadonlyArray<{ bookId: string; state: BookStateJson; coverUrl?: string; finished?: boolean; hidden?: boolean }>,
  since?: string,
): SyncManifestIndex {
  const rows: SyncManifestIndexBook[] = [];
  for (const { bookId, state, coverUrl, finished, hidden } of books) {
    const updatedAt = bookManifestUpdatedAt(state);
    if (since && updatedAt <= since) continue;
    rows.push({
      bookId,
      updatedAt,
      title: state.title,
      author: state.author,
      series: state.series,
      seriesPosition: state.seriesPosition,
      chapterCount: state.chapters.filter((c) => !c.excluded).length,
      ...(coverUrl ? { coverUrl } : {}),
      ...(finished ? { finished: true } : {}),
      ...(hidden ? { hidden: true } : {}),
    });
  }
  return {
    schemaVersion: SYNC_MANIFEST_SCHEMA,
    books: rows,
    // Full set regardless of `since` — drives client-side eviction.
    activeBookIds: books.map((b) => b.bookId),
  };
}

export function buildSyncManifestBookDetail(
  bookId: string,
  state: BookStateJson,
  audioByChapterId: ReadonlyMap<number, ChapterAudioFact>,
): SyncManifestBookDetail {
  const chapters: SyncManifestChapter[] = [];
  for (const c of state.chapters) {
    if (c.excluded) continue;
    const audio = audioByChapterId.get(c.id);
    const lufs = c.audioQa?.measuredLufs;
    const durationSec = audio?.durationSec ?? c.audioQa?.durationSec;
    chapters.push({
      // The route guarantees a uuid via ensureChapterUuids before building.
      uuid: c.uuid ?? '',
      id: c.id,
      title: c.title,
      ...(audio
        ? {
            fingerprint: chapterFingerprint(c.audioRenderedAt, audio.fileSize),
            urlSuffix: audio.urlSuffix,
            audioUrl: `/api/books/${bookId}/chapters/${c.id}/${audio.urlSuffix}`,
          }
        : {}),
      ...(durationSec !== undefined ? { durationSec } : {}),
      ...(lufs !== undefined && lufs !== null ? { lufs } : {}),
    });
  }
  return {
    schemaVersion: SYNC_MANIFEST_SCHEMA,
    bookId,
    updatedAt: bookManifestUpdatedAt(state),
    chapters,
    activeChapterUuids: chapters.map((c) => c.uuid),
  };
}
